app/session.js

```diff
@@ -21,8 +21,60 @@ try {
 
 const envFromConfig = config.getConfig().env || {};
 
+// Max duration to batch session data before sending it to the renderer process.
+const BATCH_DURATION_MS = 16;
+
+// Max size of a session data batch. Note that this value can be exceeded by ~4k
+// (chunk sizes seem to be 4k at the most)
+const BATCH_MAX_SIZE = 200 * 1024;
+
+// Data coming from the pty is sent to the renderer process for further
+// vt parsing and rendering. This class batches data to minimize the number of
+// IPC calls. It also reduces GC pressure and CPU cost: each chunk is prefixed
+// with the window ID which is then stripped on the renderer process and this
+// overhead is reduced with batching.
+class DataBatcher extends EventEmitter {
+  constructor(uid) {
+    super();
+    this.uid = uid;
+    this.decoder = new StringDecoder('utf8');
```

Comment:
UTF8? hyper is already using binary transport for the pty? We might add this eventually to xterm.js in the future, had some good results with it lifting some burden from the input parsing.

Reply:
Not sure I follow this one. IIUC electron can't send `Buffer` objects to the renderer process so we must turn the input into utf8. If we move parsing to a worker, we might be able to use `SharedArrayBuffer` to minimize conversions

Reply:
But then why are you dealing with UTF8 at all? node-pty is supposed to return JS strings already, if encoding is omitted (defaults to UTF8 being decoded by the underlying `Socket`).

So how do you send data over to the window process? This is not able to handle binary data?

Reply:
> but gets BASE64 encoded and a JSON envelope

Exactly :(

> falling back to websockets

IPC doesn't seem to cost that much though, V8 strings are pretty fast. In this profile of one frame, IPC only takes a tiny sliver (<0.5ms), not a lot of gains there. It might become more prominent when we do WebWorkers though, with an extra hop, but then again, we'll probably use `SharedArrayBuffer` for that

![image](https://user-images.githubusercontent.com/1410520/50102709-bbbe4380-01f3-11e9-9113-fc715eff33f8.png)

Reply:
I just realized that real gains are hidden in the parsing of strings part (i.e. your fix https://github.com/xtermjs/xterm.js/pull/1796), so ignore that last part :sweat_smile:

Reply:
Yeah thats the basic idea behind my optimization plan (#1811) - by getting of any string handling in the input chain and working directly on arraybuffer we can lower the runtime needed for input handling by >50%.
This gets even more if we change to direct utf8 binary input as it would save the nonsense conversions. If you are interested I have a heavily optimized playground branch, where you can test it in the demo app (https://github.com/jerch/xterm.js/tree/tttt, also needs https://github.com/jerch/node-textdecode)

Reply:
If you cannot get it run, here are my results: https://github.com/xtermjs/xterm.js/issues/791#issuecomment-447993547 :smile: